scipy/linalg/_special_matrices.py

```diff
@@ -284,7 +284,7 @@ def hadamard(n, dtype=int):
     return H
 
 
-@_apply_over_batch(("f", 1), ("s", 1))
+@_apply_over_batch(("f", 1), ("s", 1), signature="(i),(j)->(i,i)")
```

Comment:
This is one of those cases where it would be nice to be able to do something like `signature="(i),(i-1)->(i,i)"`. Right now, `j` can be anything and the decorator wouldn't complain. It would be better if the decorator verified that the second argument is `i-1`, where `i` has already been determined from the first argument. I think I'll add this capability.

It would also be nice to be able to add more general constraints on the values that really aren't part of the NEP 5 syntax. For instance, `signature="(i),(i-1>0)->(i,i)"` would check that the second argument is at least 1, and if not, raise as it does for the case when the batch size is nonzero. (Then again, maybe neither of these cases should raise, and the result should just be size-zero array with the correct shape.)